### PR TITLE
added_g_covariance

### LINKED ISSUE #72: Add function `g_covariance()`
depends on #71 

To do:

- [x] See https://github.com/insightsengineering/tern.mmrm/blob/13262696496822792e593f691dfb2574f7d45335/design/covariance_plot.Rmd#L105
- [x] Add function `g_covariance()`
  - [x] consider comment https://github.com/insightsengineering/tern.mmrm/pull/67/files#r903376719
- [ ] Add unit tests
- [x] Add roxygen2 documentation
- [x] Add to `pkgdown.yaml`

### PR DESCRIPTION
**Pull request by juliadedic1**
closes #72 however, unit tests still outstanding

### CODE REVIEW

**Comment by juliadedic1** [comment]
Hi @danielinteractive I'm pretty much done adding the `g_covariance` function except for the unit tests. I was wondering if you had any advice for creating unit test for graph results. I considered `expect_identical(result, expected)` as seen in the simple test.

File: R/covariance_plot.R [hunk 1]
```diff
@@ -77,11 +76,49 @@ h_vectorization <- function(vcov_matrix, string = NULL) {
...
+  rank_col <- as.numeric(as.factor(timepoints$col_time))
   lag <- rank_col - rank_row
   if (any(is.na(lag)) | any(is.na(time_point_distribution))) {
     warning("Verify you have used the string argument correctly or
   that the columns and rows of your input matrix are named as expected")
   }
   data.frame(Vect = vect, time_point_distribution, lag, rank_row, rank_col)
 }
+#' Function to graph covariance
+#'
```
Comment (danielinteractive) [inline R/covariance_plot.R:88]: this function is exported therefore please add lifecycle badge (experimental)

File: R/covariance_plot.R [hunk 2]
```diff
@@ -77,11 +76,49 @@ h_vectorization <- function(vcov_matrix, string = NULL) {
...
+#' @description Plot of covariance or correlation structures as a function of lag or time. The covariance structure
```
Comment (danielinteractive) [inline R/covariance_plot.R:89]: the badge should then come after the description tag, see other R files for examples

File: R/covariance_plot.R [hunk 3]
```diff
@@ -77,11 +76,49 @@ h_vectorization <- function(vcov_matrix, string = NULL) {
...
+#' needs to be vectorized and lag or time distances computed
+#' @param vcov_matrix (`matrix`)\cr name of the input symmetric matrix.
```
Comment (danielinteractive) [inline R/covariance_plot.R:91]:
```suggestion
#' 
#' @param vcov_matrix (`matrix`)\cr name of the input symmetric matrix.
```
(easier to parse)

File: R/covariance_plot.R [hunk 4]
```diff
@@ -77,11 +76,49 @@ h_vectorization <- function(vcov_matrix, string = NULL) {
...
+#' @param x_var (`string`)\cr can be "lag" or "time_point_distribution" for lag and time distance
+#' respectively. The default value is "lag".
+#' @param legend_position (`string`)\cr denoting where the legend should be shown. Option
+#' are c("topright","topleft","bottomright","bottomleft"). Default #' value #' is "topright"
```
Comment (danielinteractive) [inline R/covariance_plot.R:95]:
```suggestion
#' @param legend_position (`string`)\cr denoting where the legend should be shown.
```
no need in general to repeat arguments in the documentation, as the user sees that in the help page already, same for defaults

File: R/covariance_plot.R (hunk 3 above)
Comment (danielinteractive) [inline R/covariance_plot.R:91]:
```suggestion
#' @param vcov_matrix (`matrix`)\cr covariance matrix.
```
easier

File: R/covariance_plot.R [hunk 5]
```diff
@@ -77,11 +76,49 @@ h_vectorization <- function(vcov_matrix, string = NULL) {
...
   }
   data.frame(Vect = vect, time_point_distribution, lag, rank_row, rank_col)
 }
+#' Function to graph covariance
+#'
+#' @description Plot of covariance or correlation structures as a function of lag or time. The covariance structure
+#' needs to be vectorized and lag or time distances computed
+#' @param vcov_matrix (`matrix`)\cr name of the input symmetric matrix.
+#' @param x_var (`string`)\cr can be "lag" or "time_point_distribution" for lag and time distance
+#' respectively. The default value is "lag".
```
Comment (danielinteractive) [inline R/covariance_plot.R:93]:
```suggestion
#' respectively.
```

File: R/covariance_plot.R [hunk 6]
```diff
@@ -77,11 +76,49 @@ h_vectorization <- function(vcov_matrix, string = NULL) {
...
+#' @param legend_position (`string`)\cr denoting where the legend should be shown. Option
+#' are c("topright","topleft","bottomright","bottomleft"). Default #' value #' is "topright"
+#' @examples
+#' g_covariance(vcov_matrix, x_var = "time_point_distribution")
+g_covariance <- function(vcov_matrix, string = NULL, x_var = c("lag", "time_point_distribution"), ylab = "",
+                         xlab = NULL, col = tre_col, pch = NULL, lty = NULL, cex = 2,
+                         legend_position = c("topright", "topleft", "bottomright", "bottomleft"), ...) {
```
Comment (danielinteractive) [inline R/covariance_plot.R:100]: please put every argument on its own line, then easier to parse

File: R/covariance_plot.R [hunk 7]
```diff
@@ -77,11 +76,49 @@ h_vectorization <- function(vcov_matrix, string = NULL) {
...
+    "#0080ff", "#ff00ff", "darkgreen", "#ff0000", "orange",
+    "#00ff00", "brown"
+  )
+  x_var <- match.arg(x_var)
+  if (x_var == "lag" & is.null(xlab)) {
+    xlab <- "Lag"
+  }
+  if (x_var == "time_point_distribution" & is.null(xlab)) {
+    xlab <- "Distance (time units) btw measurements"
+  }
```
Comment (danielinteractive) [inline R/covariance_plot.R:111]: use `xlab <- if () else {}` pattern

File: tests/testthat/test_covariance_plot.R [hunk 8]
```diff
@@ -84,3 +84,27 @@ test_that("h_vectorization works as expected when string not specified", {
...
+    time_point_distribution = c(0, 1, 0),
+    lag = c(0, 1, 0),
+    rank_row = c(1, 1, 2),
+    rank_col = c(1, 2, 2)
+  )
+  expected <- ggplot(expected_dataframe, aes(x = lag, y = Vect, colour = rank_row, group = rank_row)) +
+    geom_point() +
+    geom_line() +
+    labs(xlab = "lag", ylab = "")
+  expect_identical(result, expected)
```
Comment (danielinteractive) [inline tests/testthat/test_covariance_plot.R:109]: please use `vdiffr::expect_doppelgaenger()`, see https://github.com/insightsengineering/tern.mmrm/blob/3c569e88990a207393cd2558845b47fe09b75b68/tests/testthat/test-g_mmrm.R#L17 for an example

**Review by danielinteractive — COMMENTED** [review]
thx a lot Julia, nice, please see comments inside

File: tests/testthat/test_covariance_plot.R (hunk 8 above)
Comment (juliadedic1) [reply tests/testthat/test_covariance_plot.R:109]: Thank you @danielinteractive. I have added the unit tests but I am a little confused / curious on how this tests exactly when you plug result into `vdiffr::expect_doppelganger()`

Comment (danielinteractive) [reply tests/testthat/test_covariance_plot.R]: Honestly I don't know the internals, you can find more documentation here: https://vdiffr.r-lib.org/reference/expect_doppelganger.html

File: R/covariance_plot.R [hunk 9]
```diff
@@ -77,11 +76,59 @@ h_vectorization <- function(vcov_matrix, string = NULL) {
...
+    "#00ff00", "brown"
+  )
+  x_var <- match.arg(x_var)
+  if (x_var == "lag" & is.null(xlab)) {
+    xlab <- "Lag"
+  } else if (x_var == "time_point_distribution" & is.null(xlab)) {
+    xlab <- "Distance (time units) btw measurements"
+  } else if (!is.null(xlab)) {
+    xlab <- xlab
+  }
```
Comment (danielinteractive) [inline R/covariance_plot.R:121]:
```suggestion
  if (is.null(xlab)) {
    xlab <- if (x_var == "lag") "Lag" else "Distance (time units) btw measurements" 
  }
```
that is doing the same and is much easier readable

**Review by danielinteractive — COMMENTED** [review]
thanks Julia, please run the tests such that the snapshot files are populated in the tests directory

**Comment by juliadedic1** [comment]
Hi @danielinteractive added in those snapshots and fixed a few errors

**Review by danielinteractive — APPROVED** [review]

**Pull request merged**